### PR TITLE
close #1632 generate bib number

### LINKED ISSUE #1632: Generate BIB number

### PR DESCRIPTION
**Pull request by vaneath**
[Screencast from 2024-07-22 17-30-03.webm](https://github.com/user-attachments/assets/8c9a511d-6ddd-4817-9a52-3c340f04fb4e)

### CODE REVIEW

File: app/jobs/spree_cm_commissioner/generate_bib_number_job.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+module SpreeCmCommissioner
+  class GenerateBibNumberJob < ApplicationJob
+    def perform(taxon_id)
+      taxon = Spree::Taxon.find(taxon_id)
```
Comment (channainfo) [inline app/jobs/spree_cm_commissioner/generate_bib_number_job.rb:4]: Move the logic to BibNumberGenerator class and rename this to BibNumberGeneratorJob

**Review by channainfo — CHANGES_REQUESTED** [review]

**Comment by channainfo** [comment]
The bib number and the trail should be stored as meta fields in the line items. Storing the bib in the guest is not ideal because not all products require a bib. I think we might not need the bib number auto generated as well.

File: app/models/concerns/spree_cm_commissioner/bib_number_concern.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module SpreeCmCommissioner
+  module BibNumberConcern
+    extend ActiveSupport::Concern
+
+    included do
+      after_update :generate_bib_number_job, if: -> { state_changed_to_complete? }
```
Comment (channainfo) [inline app/models/concerns/spree_cm_commissioner/bib_number_concern.rb:6]: Let's generate bib number only if the product bib prefix is present

File: db/migrate/20240726043406_add_partial_unique_index_to_bib_number_and_bib_number_prefix.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class AddPartialUniqueIndexToBibNumberAndBibNumberPrefix < ActiveRecord::Migration[7.0]
+  def change
+    add_index :cm_guests, [:bib_number, :bib_number_prefix, :event_id],
```
Comment (channainfo) [inline db/migrate/20240726043406_add_partial_unique_index_to_bib_number_and_bib_number_prefix.rb:3]: Let's remove this for now. Create a new ticket for this. The code here is complicated

**Comment by theachoem** [comment]
If we used this https://github.com/channainfo/commissioner/pull/1705 instead, let's close this PR as duplicate.

**Comment by vaneath** [comment]
Move to #1705

**Pull request closed without merging**